l1-contracts/contracts/bridge/asset-router/L1AssetRouter.sol

```diff
@@ -0,0 +1,582 @@
+
+    /// @inheritdoc IAssetRouterBase
+    function finalizeDeposit(
+        uint256 _chainId,
+        bytes32 _assetId,
+        bytes calldata _transferData
+    ) public override(AssetRouterBase, IAssetRouterBase) onlyNullifier {
+        _finalizeDeposit(_chainId, _assetId, _transferData, address(nativeTokenVault));
+        emit DepositFinalizedAssetRouter(_chainId, _assetId, _transferData);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            CLAIM FAILED DEPOSIT Functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IL1AssetRouter
+    function bridgeRecoverFailedTransfer(
+        uint256 _chainId,
+        address _depositSender,
+        bytes32 _assetId,
+        bytes calldata _assetData
+    ) external override onlyNullifier nonReentrant whenNotPaused {
+        IL1AssetHandler(assetHandlerAddress[_assetId]).bridgeRecoverFailedTransfer(
+            _chainId,
+            _assetId,
+            _depositSender,
+            _assetData
+        );
+
+        emit ClaimedFailedDepositAssetRouter(_chainId, _assetId, _assetData);
+    }
+    /*//////////////////////////////////////////////////////////////
+                     Internal & Helpers
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Decodes the transfer input for legacy data and transfers allowance to NTV.
+    /// @dev Is not applicable for custom asset handlers.
+    /// @param _data The encoded transfer data (address _l1Token, uint256 _depositAmount, address _l2Receiver).
+    // / @param _prevMsgSender The address of the deposit initiator.
+    /// @return Tuple of asset ID and encoded transfer data to conform with new encoding standard.
+    function _handleLegacyData(bytes calldata _data, address) internal override returns (bytes32, bytes memory) {
+        (address _l1Token, uint256 _depositAmount, address _l2Receiver) = abi.decode(
+            _data,
+            (address, uint256, address)
+        );
+        bytes32 assetId = _ensureTokenRegisteredWithNTV(_l1Token);
+        // L1_NULLIFIER.transferAllowanceToNTV(assetId, _depositAmount, _prevMsgSender);
+        return (assetId, abi.encode(_depositAmount, _l2Receiver));
+    }
+
+    /// @notice Ensures that token is registered with native token vault.
+    /// @dev Only used when deposit is made with legacy data encoding format.
+    /// @param _token The L1 token address which should be registered with native token vault.
+    /// @return assetId The asset ID of the token provided.
+    function _ensureTokenRegisteredWithNTV(address _token) internal returns (bytes32 assetId) {
+        assetId = nativeTokenVault.getAssetId(block.chainid, _token);
+        if (nativeTokenVault.tokenAddress(assetId) == address(0)) {
+            nativeTokenVault.registerToken(_token);
+        }
+    }
+
+    /// @inheritdoc IL1AssetRouter
+    function transferAllowanceToNTV(
```

Comment:
maybe I am wrong, but it seems to me that the order of calls is the following:

L1NTV._bridgeBurnNativeToken -> L1_ASSET_ROUTER.transferAllowanceToNTV

All the `L1_ASSET_ROUTER.transferAllowanceToNTV` does is transferring funds from the user to the L1AssetRouter and set the allowance for L1AssetRouter to the L1NTV.

Now, if the NTV._bridgeBurnNativeToken is used, it will transfer funds from the user (prev sender) and not the bridge. It seems like it will not work if the user approves funds to the shared bridge and not the ntv.

Reply:
True connected, we return now if deposit was checker or not